internal/util/secretskms.go

```diff
@@ -0,0 +1,227 @@
+// for encrypting/decrypting DEKs that are stored in a detached DEKStore.
+func initSecretsMetadataKMS(encryptionKMSID string, secrets map[string]string) (EncryptionKMS, error) {
+	eKMS, err := initSecretsKMS(secrets)
+	if err != nil {
+		return nil, err
+	}
+
+	sKMS, ok := eKMS.(SecretsKMS)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert %T to SecretsKMS", eKMS)
+	}
+
+	smKMS := SecretsMetadataKMS{}
+	smKMS.SecretsKMS = sKMS
+	smKMS.encryptionKMSID = encryptionKMSID
+
+	return smKMS, nil
+}
+
+// GetID is returning ID representing the SecretsMetadataKMS.
+func (kms SecretsMetadataKMS) GetID() string {
+	return kms.encryptionKMSID
+}
+
+// Destroy frees all used resources.
+func (kms SecretsMetadataKMS) Destroy() {
+	kms.SecretsKMS.Destroy()
+}
+
+func (kms SecretsMetadataKMS) requiresDEKStore() DEKStoreType {
+	return DEKStoreMetadata
+}
+
+// encryptedMetedataDEK contains the encrypted DEK and the Nonce that was used
+// during encryption. This structure is stored (in JSON format) in the DEKStore
+// that is linked to this KMS provider.
+type encryptedMetedataDEK struct {
+	// DEK is the encrypted data-encryption-key for the volume.
+	DEK []byte `json:"dek"`
+	// Nounce is a random byte slice to guarantee the uniqueness of the
+	// encrypted DEK.
+	Nonce []byte `json:"nonce"`
+}
+
+// EncryptDEK encrypts the plainDEK with a key derived from the passphrase from
+// the SecretsKMS and the volumeID.
+// The resulting encryptedDEK contains a JSON with the encrypted DEK and the
+// nonce that was used for encrypting.
+func (kms SecretsMetadataKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
+	// use the passphrase from the SecretsKMS
+	passphrase, err := kms.SecretsKMS.FetchDEK(volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to get passphrase: %w", err)
+	}
+
+	aead, err := generateCipher(passphrase, volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate cipher: %w", err)
+	}
+
+	emd := encryptedMetedataDEK{
+		Nonce: generateNonce(aead.NonceSize()),
+	}
+	emd.DEK = aead.Seal(nil, emd.Nonce, []byte(plainDEK), nil)
+
+	emdData, err := json.Marshal(&emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert "+
+			"encryptedMetedataDEK to JSON: %w", err)
+	}
+
+	return string(emdData), nil
+}
+
+// DecryptDEK takes the JSON formatted `encryptedMetedataDEK` contents, and it
+// fetches SecretsKMS passphase to decrypt the DEK.
+func (kms SecretsMetadataKMS) DecryptDEK(volumeID, encryptedDEK string) (string, error) {
+	// use the passphrase from the SecretsKMS
+	passphrase, err := kms.SecretsKMS.FetchDEK(volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to get passphrase: %w", err)
+	}
+
+	aead, err := generateCipher(passphrase, volumeID)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate cipher: %w", err)
+	}
+
+	emd := encryptedMetedataDEK{}
+	err = json.Unmarshal([]byte(encryptedDEK), &emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert data to "+
```

Comment:
same here as well. return and logging of error will leak any sensitive information?

Reply:
no, I do not think so